### PR TITLE
WIP: Segmentation

### PR DESCRIPTION
**Pull request by avv-va**

### CODE REVIEW

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -525,6 +526,144 @@ def calculate_keypoints_loss(self, masks, target_gt_idx, keypoints, batch_idx, s
...
+
+
+        anchor_intersects_gtbbox = torch.zeros((batch_size, max_detections_len, num_anchors), device=self.device, dtype=torch.bool) # (B, T, A)
+        for b in range(batch_size):
+            bbox_x1 = gt_bboxes[b, :, 0].unsqueeze(1)  # (T, 1)
+            bbox_y1 = gt_bboxes[b, :, 1].unsqueeze(1)
+            bbox_x2 = gt_bboxes[b, :, 2].unsqueeze(1)
+            bbox_y2 = gt_bboxes[b, :, 3].unsqueeze(1)
+            no_intersect_in_batch = ((ax1 > bbox_x2) | (bbox_x1 > ax2) | (ay1 > bbox_y2) | (bbox_y1 > ay2))  # (T, A)
+            anchor_intersects_gtbbox[b] = ~no_intersect_in_batch  # (T, A) True if anchor box intersects with gt box
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:651]: One line equivalent
```
anchor_intersects_gtbbox[b] = ((ax1 <= bbox_x2) & (bbox_x1 <= ax2) & (ay1 <= bbox_y2) & (bbox_y1 <= ay2))  # (T, A) True if anchor box intersects with gt box
```

File: .gitignore [hunk 2]
```diff
@@ -162,4 +162,4 @@ weights/
 pnnx*
 
 # Autogenerated files for tests
-/ultralytics/assets/
+/ultralytics/assets/
```
Comment (romainthibaux) [inline .gitignore:165]: No reason to change (removed newline)

File: training_utils/quick_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from export import Export
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Evaluate the model")
+    parser.add_argument("-l", "--load", type=str, default=None, help="Path to the model weights to load. Load the pretrained model")
+    args = parser.parse_args()
+    PrepareDataset(coco_classes_file, dataset_yaml_path, training_task)
+
+    args.load = 'runs/pose-segmentation/train82/weights/best.pt'
```
Comment (romainthibaux) [inline training_utils/quick_evaluation.py:24]: This should not be hard coded.

File: ultralytics/models/yolo/pose_seg/val.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        """
+        preds = x_flat, ([P1, P2, P3], kpt)
+        Each Pi is (bs, self.no, h_i, w_i), with h_i and w_i being different for each P. e.g 8x8, 4x4, 2x2 corresponding to resolution(self.stride)
+        x_flat = (bs, 8=xyxy(bbox),cls0,cls1,seg0,seg1, anchors_len) e.g anchors_len = 8x8 + 4x4 + 2x2 = 84
+        """
+        Pi_list = preds[1][0]  # [P1, P2, P3] Needs to be returned to reconstruct anchors later on
+        x_flat = preds[0]
+        seg_logits = x_flat[:, 6:8, :]
+        seg0 = seg_logits[:, 0, :]
+        seg1 = seg_logits[:, 1, :]
```
Comment (romainthibaux) [inline ultralytics/models/yolo/pose_seg/val.py:33]: This makes a hard coded assumption that there are 2 detection channels and 2 segmentation channels. It should be generic, using `nc`.

File: ultralytics/models/yolo/pose_seg/val.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            # Append correct_masks, correct_boxes, pconf, pcls, tcls
+            self.stats.append((correct_bboxes, correct_kpts, pred[:, 4], pred[:, 5], cls.squeeze(-1)))
+
+            # Save
+            if self.args.save_json:
+                self.pred_to_json(predn, batch['im_file'][si])
+            # if self.args.save_txt:
+            #    save_one_txt(predn, save_conf, shape, file=save_dir / 'labels' / f'{path.stem}.txt')
+
+    def map_anchors_to_seg(self, seg_logits, seg_mask, Pi_list):
```
Comment (romainthibaux) [inline ultralytics/models/yolo/pose_seg/val.py:109]: Please add a docstring explaining briefly what the function does, and what the arguments are. In particular I don't understand `seg_mask`.

File: ultralytics/models/yolo/pose_seg/val.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        strides = strides.unsqueeze(0).repeat(bs, 1, 1) # (B, 1, A)
+        anchor_points = anchor_points.unsqueeze(0).repeat(bs, 1, 1) # (B, 2, A) 
+
+        seg_results = []
+        for b in range(seg_logits.shape[0]):
+            selected_anchor_points = anchor_points[b, :, seg_mask[b]]
+            selected_strides = strides[b, :, seg_mask[b]]
+            selected_seg = seg_logits[b, :, seg_mask[b]]
+            conf_c = selected_seg[0, :] > self.args.seg_conf_c
+            conf_w = selected_seg[1, :] > self.args.seg_conf_w
```
Comment (romainthibaux) [inline ultralytics/models/yolo/pose_seg/val.py:126]: Again this hard codes that there are 2 segmentation channels.

File: ultralytics/nn/modules/head.py [hunk 7]
```diff
@@ -391,7 +391,102 @@ def kpts_decode(self, bs, kpts):
             y[:, 0::ndim] = (y[:, 0::ndim] * 2.0 + (self.anchors[0] - 0.5)) * self.strides
             y[:, 1::ndim] = (y[:, 1::ndim] * 2.0 + (self.anchors[1] - 0.5)) * self.strides
             return y
-    
+
+
+class DetectAndSeg(Detect):
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:396]: That's a lot of copy-paste. It might be inevitable but let's discuss.

File: ultralytics/utils/loss.py [hunk 8]
```diff
@@ -525,6 +525,140 @@ def calculate_keypoints_loss(self, masks, target_gt_idx, keypoints, batch_idx, s
...
+                                                             stride_tensor, target_bboxes, pred_kpts, batch['ignore_kpt'])
+
+            loss[5] = self.calculate_seg_loss(
+                pred_seg=pred_seg,
+                anchor_points=anchor_points,
+                stride_tensor=stride_tensor,
+                gt_bboxes=gt_bboxes,
+                gt_labels=gt_labels,
+                mask_gt=mask_gt
+            )
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:591]: There is no reason for this loss to only be defined when there is at least one box.

File: ultralytics/utils/plotting.py [hunk 9]
```diff
@@ -465,7 +468,32 @@ def plot_images(images,
...
+            if seg_results is not None:
+                image_segs = seg_results[i]
+                seg_center_xy = image_segs[:, :2] * (image_segs[:, 2].unsqueeze(1).repeat(1, 2))
+                strides = image_segs[:, 2].unsqueeze(1)
+                seg_class = image_segs[:, 3].unsqueeze(1)
+
+                for cx_orig, cy_orig, seg_class, stride in zip(seg_center_xy[:, 0], seg_center_xy[:, 1], seg_class, strides):
+                    if stride not in res_grid_size:
+                        continue
+                    if seg_class == -1: # to visualize segments that belong to both classes
```
Comment (romainthibaux) [inline ultralytics/utils/plotting.py:480]: Hard coded 2 classes.
